scipy/sparse/construct.py

```diff
@@ -667,16 +667,34 @@ def block_diag(mats, format=None, dtype=None):
            [0, 0, 0, 7]])
 
     """
-    nmat = len(mats)
-    rows = []
-    for ia, a in enumerate(mats):
-        row = [None]*nmat
+    row = []
+    col = []
+    data = []
+    r_idx = 0
+    c_idx = 0
+    for a in mats:
+        if isinstance(a, (list, numbers.Number)):
+            a = coo_matrix(a)
+        nrows, ncols = a.shape
         if issparse(a):
-            row[ia] = a
+            a = a.tocoo()
+            row.append(a.row + r_idx)
+            col.append(a.col + c_idx)
+            data.append(a.data)
         else:
-            row[ia] = coo_matrix(a)
-        rows.append(row)
-    return bmat(rows, format=format, dtype=dtype)
+            a_row = np.repeat(np.arange(r_idx, r_idx + nrows), ncols)
+            a_col = np.tile(np.arange(c_idx, c_idx + ncols), nrows)
+            row.append(a_row + r_idx)
+            col.append(a_col + c_idx)
```

Comment:
These two lines now need the row and column offsets removed, since they're included included in the new direct constructors above:
```
row.append(a_row)
col.append(a_col)
```

Reply:
Actually I suppose we don't need any local references to `a_row` and `a_col` then, and it can all be combined to just
```
row.append(np.repeat(np.arange(r_idx, r_idx + nrows), ncols))
col.append(np.tile(np.arange(c_idx, c_idx + ncols), nrows))
```

Reply:
I messed with the commit and have reverted this change.

Reply:
@kburns: I think that keepings those lines makes it clearer.